File: src/main/kotlin/com/huawei/cangjie/psi/stubs/elements/CjMainFunctionElementType.java
package com.huawei.cangjie.psi.stubs.elements;

import com.huawei.cangjie.name.FqName;
import com.huawei.cangjie.psi.CjFile;
import com.huawei.cangjie.psi.CjMainFunction;
import com.huawei.cangjie.psi.psiUtil.CjPsiUtilKt;
import com.huawei.cangjie.psi.stubs.CangJieFunctionStub;

import com.huawei.cangjie.psi.stubs.impl.CangJieFunctionStubImpl;
import com.huawei.cangjie.psi.stubs.impl.CangJieStubOrigin;
import com.intellij.psi.PsiElement;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import com.intellij.util.io.StringRef;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;


public class CjMainFunctionElementType extends CjStubElementType<CangJieFunctionStub, CjMainFunction> {
    private static final String NAME = "cangjie.MAIN_FUNCTION";

    public CjMainFunctionElementType(@NotNull String debugName, @NotNull Class<CjMainFunction> psiClass, @NotNull Class<?> stubClass) {
        super(debugName, psiClass, stubClass);
    }


    public CjMainFunctionElementType(@NotNull String debugName) {
        super(debugName, CjMainFunction.class, CangJieFunctionStub.class);
    }


    @Override
    public @NotNull CangJieFunctionStubImpl createStub(@NotNull CjMainFunction psi, StubElement<? extends PsiElement> parentStub) {
        boolean isTopLevel = psi.getParent() instanceof CjFile;
        boolean isExtension = psi.getReceiverTypeReference() != null;
        FqName fqName = CjPsiUtilKt.safeFqNameForLazyResolve(psi);
        boolean hasBlockBody = psi.hasBlockBody();
        boolean hasBody = psi.hasBody();
        return new CangJieFunctionStubImpl(
                parentStub,CjStubElementTypes.MAIN_FUNC, StringRef.fromString(psi.getName()), isTopLevel, fqName,
                isExtension, hasBlockBody, hasBody, psi.hasTypeParameterListBeforeFunctionName(),

null,
             null
        );
    }



    @Override
    public void serialize(@NotNull CangJieFunctionStub stub, @NotNull StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());
        dataStream.writeBoolean(stub.isTopLevel());

        FqName fqName = stub.getFqName();
        dataStream.writeName(fqName != null ? fqName.asString() : null);

        dataStream.writeBoolean(stub.isExtension());
        dataStream.writeBoolean(stub.hasBlockBody());
        dataStream.writeBoolean(stub.hasBody());
        dataStream.writeBoolean(stub.hasTypeParameterListBeforeFunctionName());
//        boolean haveContract = stub.mayHaveContract();
//        dataStream.writeBoolean(haveContract);
        if (stub instanceof CangJieFunctionStubImpl stubImpl) {


            CangJieStubOrigin.serialize(stubImpl.getOrigin(), dataStream);
        }
    }

    @Override
    public @NotNull CangJieFunctionStubImpl deserialize(@NotNull StubInputStream dataStream, StubElement parentStub) throws IOException {
        StringRef name = dataStream.readName();
        boolean isTopLevel = dataStream.readBoolean();

        StringRef fqNameAsString = dataStream.readName();
        FqName fqName = fqNameAsString != null ? new FqName(fqNameAsString.toString()) : null;

        boolean isExtension = dataStream.readBoolean();
        boolean hasBlockBody = dataStream.readBoolean();
        boolean hasBody = dataStream.readBoolean();
        boolean hasTypeParameterListBeforeFunctionName = dataStream.readBoolean();
//        boolean mayHaveContract = dataStream.readBoolean();
        return new CangJieFunctionStubImpl(
                (StubElement<?>) parentStub,CjStubElementTypes.MAIN_FUNC, name, isTopLevel, fqName, isExtension, hasBlockBody, hasBody,
                hasTypeParameterListBeforeFunctionName,
null,
                CangJieStubOrigin.deserialize(dataStream)
        );
    }
}

File: psi/src/main/kotlin/org/cangnova/cangjie/psi/stubs/elements/CjMainFunctionElementType.kt
/*
 * Copyright 2026 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */
package org.cangnova.cangjie.psi.stubs.elements

import org.cangnova.cangjie.builtins.StandardNames.MAIN
import org.cangnova.cangjie.name.FqName
import org.cangnova.cangjie.psi.CjFile
import org.cangnova.cangjie.psi.CjMainFunction
import org.cangnova.cangjie.psi.psiUtil.safeFqNameForLazyResolve
import org.cangnova.cangjie.psi.stubs.CangJieMainFunctionStub
import org.cangnova.cangjie.psi.stubs.elements.StubIndexService.Companion.getInstance
import org.cangnova.cangjie.psi.stubs.impl.CangJieMainFunctionStubImpl
import org.cangnova.cangjie.psi.stubs.impl.CangJieStubOrigin.Companion.deserialize
import org.cangnova.cangjie.psi.stubs.impl.CangJieStubOrigin.Companion.serialize
import com.intellij.psi.PsiElement
import com.intellij.psi.stubs.IndexSink
import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import com.intellij.util.io.StringRef
import java.io.IOException

class CjMainFunctionElementType : CjStubElementType<CangJieMainFunctionStub, CjMainFunction> {
    constructor(debugName: String, psiClass: Class<CjMainFunction>, stubClass: Class<*>) : super(
        debugName,
        psiClass,
        stubClass,
    )

    constructor(debugName: String) : super(debugName, CjMainFunction::class.java, CangJieMainFunctionStub::class.java)

    override fun createStub(psi: CjMainFunction, parentStub: StubElement<out PsiElement>): CangJieMainFunctionStubImpl {
        var fqName = psi.safeFqNameForLazyResolve()
        if (fqName != null) {
            val firstSegment = fqName.firstSegment()
            if (firstSegment != null) {
                fqName = FqName(firstSegment.asString()).child(MAIN)
            }
        }
        return CangJieMainFunctionStubImpl(
            parentStub, CjStubElementTypes.MAIN_FUNC, StringRef.fromString(psi.name), fqName,
            null,
        )
    }

    @Throws(IOException::class)
    override fun serialize(stub: CangJieMainFunctionStub, dataStream: StubOutputStream) {
        dataStream.writeName(stub.name)

        val fqName = stub.getFqName()
        dataStream.writeName(fqName?.asString())

        if (stub is CangJieMainFunctionStubImpl) {
            serialize(stub.origin, dataStream)
        }
    }

    override fun indexStub(stub: CangJieMainFunctionStub, sink: IndexSink) {
        getInstance().indexMainFunction(stub, sink)
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>): CangJieMainFunctionStubImpl {
        val name = dataStream.readName()

        val fqNameAsString = dataStream.readName()
        val fqName = if (fqNameAsString != null) FqName(fqNameAsString.toString()) else null

        return CangJieMainFunctionStubImpl(
            parentStub, CjStubElementTypes.MAIN_FUNC, name, fqName,
            deserialize(dataStream),
        )
    }

    companion object {
        private const val NAME = "cangjie.MAIN_FUNCTION"
    }
}
